File: data_structures/linked_lists/singly_linked_lists/singly_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.length = 0

    def _is_empty(self):
        return self.head is None

    def prepend(self, value):
        node = Node(value)
        node.next = self.head
        self.head = node
        self.length += 1
        return self

    def append(self, value):
        node = Node(value)
        if self._is_empty():
            self.head = node
        else:
            current_node = self.head
            while current_node.next is not None:
                current_node = current_node.next
            current_node.next = node
        self.length += 1
        return self

    def insert(self, index, value):
        if index < 0:
            return self
        elif index >= self.length:
            return self.append(value)
        elif index == 0:
            return self.prepend(value)
        new_node = Node(value)
        if self._is_empty():
            self.head = new_node
        else:
            left_node = self._get_node_at_index(index - 1)
            right_node = left_node.next
            left_node.next = new_node
            new_node.next = right_node
        self.length += 1
        return self

    def delete(self, index):
        if index < 0 or index >= self.length:
            return self
        if self._is_empty():
            return self
        if index == 0:
            self.head = self.head.next
        else:
            left_node = self._get_node_at_index(index - 1)
            node_to_delete = left_node.next
            left_node.next = node_to_delete.next
        self.length -= 1
        return self

    def _get_node_at_index(self, index):
        position = 0
        current_node = self.head
        while position < index:
            current_node = current_node.next
            position += 1
        return current_node

    def print_as_list(self):
        array = []
        current_node = self.head
        while current_node is not None:
            array.append(current_node.value)
            current_node = current_node.next
        print(array)
    
    def reverse(self):
        if self.length == 1 or self._is_empty():
            return self
        current_node = self.head
        previous_node = None
        while current_node is not None:
            next_node = current_node.next
            current_node.next = previous_node
            previous_node = current_node
            current_node = next_node
        self.head = previous_node
        return self

File: data_structures/linked_lists/singly_linked_lists/singly_linked_list.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}


class LinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    _isEmpty() {
        return this.head === null && this.tail === null;
    }

    prepend(value) {
        let node = new Node(value);
        if (this._isEmpty()) {
            this.tail = node;
        }
        node.next = this.head;
        this.head = node;
        this.length++;
        return this;
    }

    append(value) {
        let node = new Node(value);
        if (this._isEmpty()) {
            this.head = node;
            this.tail = node;
        }
        else {
            this.tail.next = node;
            this.tail = node;
        }
        this.length++;
        return this;
    }

    insert(index, value) {
        if (index < 0) {
            return this;
        }
        else if (index === 0) {
            return this.prepend(value);
        }
        else if (index >= this.length) {
            return this.append(value);
        }
        let newNode = new Node(value);
        if (this._isEmpty()) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            let leftNode = this._getNodeAtIndex(index - 1);
            let rightNode = leftNode.next;
            leftNode.next = newNode;
            newNode.next = rightNode;
        }
        this.length++;
        return this;
    }

    delete(index) {
        if (index < 0 || index >= this.length) {
            return this
        };
        if (this._isEmpty()) {
            return this;
        }
        else if (index === 0) {
            this.head = this.head.next;
        }
        else {
            let leftNode = this._getNodeAtIndex(index - 1);
            let nodeToBeDeleted = leftNode.next;
            leftNode.next = nodeToBeDeleted.next;
            if (nodeToBeDeleted === this.tail) {
                this.tail = leftNode;
            }
        }
        this.length--;
        return this;
    }

    _getNodeAtIndex(index) {
        let currentNode = this.head;
        let position = 0;
        while (position < index) {
            currentNode = currentNode.next;
            position++;
        }
        return currentNode;
    }

    printAsList() {
        const array = [];
        let currentNode = this.head;
        while (currentNode !== null) {
            array.push(currentNode.value)
            currentNode = currentNode.next
        }
        console.log(array);
    }

    reverse() {
        if (this.length === 1 || this._isEmpty()) {
            return this;
        }
        let currentNode = this.head;
        let previousNode = null;
        while (currentNode !== null) {
            let nextNode = currentNode.next;
            currentNode.next = previousNode;
            previousNode = currentNode;
            currentNode = nextNode;
        }
        this.tail = this.head;
        this.head = previousNode;
        return this;
    }
}